add a new 'links' field to matched citations

JoshData opened the pull request:
This PR adds a new `links` field to matched citations with information to generate a permalink. I've implemented this for US Code citations as:

	$ cite "40 U.S.C. § 11101(1)"
	{
	  "citations": [
	    {
	      "type": "usc",
	      "match": "40 U.S.C. § 11101(1)",
	      "index": 0,
	      "usc": {
	        "title": "40",
	        "section": "11101",
	        "subsections": [
	          "1"
	        ],
	        "id": "usc/40/11101/1",
	        "links": [
	          {
	            "source": "U.S. Government Publishing Office",
	            "source_abbrev": "US GPO",
	            "authoritative": true,
	            "note": "Sub-section citation is not included.",
	            "links": [
	              {
	                "url": "http://api.fdsys.gov/link?collection=uscode&year=2013&title=40&section=11101&type=usc",
	                "type": "pdf"
	              },
	              {
	                "url": "http://api.fdsys.gov/link?collection=uscode&year=2013&title=40&section=11101&type=usc&link-type=html",
	                "type": "html"
	              },
	              {
	                "url": "http://api.fdsys.gov/link?collection=uscode&year=2013&title=40&section=11101&type=usc&link-type=contentdetail",
	                "type": "other"
	              }
	            ]
	          },
	          {
	            "source": "Cornell Legal Information Institute",
	            "source_abbrev": "Cornell LII",
	            "authoritative": false,
	            "links": [
	              {
	                "url": "http://www.law.cornell.edu/uscode/text/40/11101#1"
	              }
	            ]
	          }
	        ]
	      }
	    }
	  ]
	}

The `links` array is a list of sources (e.g. US GPO, Cornell LII), each of which can offer one or more links in different formats (PDF, text, etc.).

I've left in an unused argument where we'd have to consider point-in-time issues: to create a permanent link to X USC Y as it is today, or as it was on a certain past date. The output links should probably also include information about their as-of date: for US GPO links, it would be the US Code edition date, but for Cornell LII links it would be a token indicating it's most-recently-published, since the LII URLs will always reflect the most recent publication they've updated to.

What do you all think?

This only took a few minutes to cook up, so no worries if you all think this should go in a separate module perhaps.

adelevie commented:
I like this.

konklone commented:
> I've left in an unused argument where we'd have to consider point-in-time issues: to create a permanent link to X USC Y as it is today, or as it was on a certain past date. 

My gut reaction is that that's outside of the scope of this library. If the point-in-time (e.g. edition) is relevant, it should be detectable from the cite text, and part of the generated permalink -- just like every other aspect of the cite (e.g. the Title).

> I like this.

I do too! The format could be compacted though, I think. From the perspective of an integrator, I'd want to reference things by clear predictable keys, like `cite.links.gpo.url` or `cite.links.authoritative.pdf`, rather than have nested arrays that require the user to iterate through each array level looking for the object with the key/value pair they want.

I don't see a huge need to provide multiple formats of each link, either. We can be somewhat opinionated here, and if people really want the PDF instead of the HTML, they can write their own function for it.

Other thoughts?

JoshData commented:
> If the point-in-time (e.g. edition) is relevant, it should be detectable from the cite text

This is almost never true in practice, e.g. in USC citations in bills. The as-of date is implicit and, to the extent meaningful, it's the date the bill was written. But to generate a URL, a publication edition will often be required (or should be).

> rather than have nested arrays

I did this so we could have a priority order, e.g. US GPO first, Cornell last. In the future, there might be uscode.house.gov first, then gpo.gov, then Cornell. I had a particular use case/UI in mind. But I can take that out and make them all keyed by source abbreviations (usgpo, cornell_lii).

> I don't see a huge need to provide multiple formats of each link, either. We can be somewhat opinionated here, and if people really want the PDF instead of the HTML, they can write their own function for it.

If it's `cite.links.usgpo.pdf`, then it's straightforward to offer multiple formats, right? That format/path sounds great to me.

JoshData commented:
Oh wait, if it's `cite.links.usgpo.pdf`, then is there no additional link metadata, or where does it hang?

How's this?

	"links": {
	  "usgpo": {
	    "_source": {
	      "name": "U.S. Government Publishing Office",
	      "abbrev": "US GPO",
	      "authoritative": true,
	      "note": "Sub-section citation is not included."
	    },
	    "pdf": "http://api.fdsys.gov/link?collection=uscode&year=2013&title=40&section=11101&type=usc",
	    "html": "http://api.fdsys.gov/link?collection=uscode&year=2013&title=40&section=11101&type=usc&link-type=html",
	    "browse": "http://api.fdsys.gov/link?collection=uscode&year=2013&title=40&section=11101&type=usc&link-type=contentdetail"
	  },
	  "cornell_lii": {
	    "_source": {
	      "name": "Cornell Legal Information Institute",
	      "abbrev": "Cornell LII",
	      "authoritative": false
	    },
	    "html": "http://www.law.cornell.edu/uscode/text/40/11101#1"
	  }
	}

konklone commented:
> This is almost never true in practice, e.g. in USC citations in bills. The as-of date is implicit and, to the extent meaningful, it's the date the bill was written. But to generate a URL, a publication edition will often be required (or should be).

It sounds like you're saying that government lawyers make informatically inefficient choices!

I think the closest-to-right solution is to take advantage of the ability to send in [citator-specific metadata](https://github.com/unitedstates/citation#cite-specific-options), which can be passed to the `links` function so that it's available during link construction. So, if your application-specific context has access to the bill's year (for this example, in in a `bill_year` variable), you can do `Citation.find(string, {type: "usc", usc: {year: bill_year}})` and then every part of the USC citator (from the cite parser to the link constructor) has access to that additional known context.

JoshData commented:
> I think the closest-to-right solution is to take advantage of the ability to send in citator-specific metadata, 

I hadn't thought of that. But it's not really citator specific. When it's implicit, it's really just a date associated with the document. Let's maybe punt this particular question for now and just let this work for "current" citations?

JoshData commented:
PR updated (force push).

anseljh commented:
> This only took a few minutes to cook up, so no worries if you all think this should go in a separate module perhaps.

I like the idea of a separate module for this. I can think of use cases where you are generating links, but not extracting them. I **don't** feel strongly about it, though. It may not be worth the extra work, and it could always be done later.

Nice PR, @JoshData!

konklone commented:
I just made some tweaks in https://github.com/unitedstates/citation/commit/67fe99fb3419d213f855a6180cbce805e28a6546 to the field names and values, mostly aesthetic (`_source` to `source`, `http://gpo.gov/` to `http://www.gpo.gov`, `abbrev` to `abbreviation`), and minorly cleaned up the code.

@JoshData, unfortunately this did break 9 tests, which were checking for the full equality of the resulting data object. I think the tests should be a little more forgiving of changes unrelated to what they are testing, but if these link functions are going in core and not a module, they should have some tests. The `cfr.js` function was broken and returning incomplete URLs because of a missing `+` sign.

I think there's a pretty strong case for moving these to a module, but I'm happy accepting it to core for the time being. However, I do think the `links` field should be opt in, as it's a major increase of bandwidth and storage for response sizes. At scale (like the system powering [Scout](https://scout.sunlightfoundation.com/search/all/5%20usc%20552)), that can matter.

I hate to spend so much time going back and forth on this stuff, especially since you've got it running somewhere, but I think it's worth it for a very nice feature like this. Can you fix the tests and add at least a couple for this feature? If the `nodeunit` tests are difficult to understand, happy to help adapt any test cases you can define.

JoshData commented:
Great. I made it optional via `--links`/`links: true` -- which has the nice effect of making it pass tests.

I also added documentation in the README and changed the resource type for Cornell LII pages from html to landing. We can use html to refer to getting back raw HTML for the law itself (rather than a web page about it).

I'll come back in a separate commit for new tests.

JoshData commented:
I added a bunch of tests and links for more of the citators. Good to merge?

konklone commented:
Looks good, rebase and merge away! Thanks, @joshdata.

JoshData commented:
Thanks everyone!
